### ISSUE TITLE
Ring CurlHandler showing as culprit for huge delays in New Relic - misnomer?

### ISSUE THREAD

**Issue by cpats007** [guzzle/guzzle#1690]
Hi All.

We're currently using Guzzle v6 across a lot of our applications which are being monitored with New Relic. In certain instances, when there are huge delays in response times, the trace in New Relic is showing the following as the culprit, taking 90%+ of the time:

`GuzzleHttp\Ring\Client\CurlHandler::_invokeAsArray`

My initial suspicion is that this isn't really the case, and it just happens to be pausing on that trace for some reason, but can someone shed light on what this functionality does, and whether it is possible that this could cause massive delays in certain instances (by massive, I'm talking from a normal application response time of around 40ms, to over 500ms and in some instances this is being reported as taking up to 40 / 50 seconds!)

An idea of what New Relic is showing is here:

![image](https://cloud.githubusercontent.com/assets/5278964/21256397/ea05d39a-c369-11e6-95a6-563b90161153.png)

If you need any more information I can provide it, but I wanted to check if this may be a common / diagnosed issue and whether there was anything I could do to either code around it, or resolve this possible issue. If I understood what this function did, potentially I could try and circumnavigate it and test if this is indeed an issue or not.

Thanks all

**Comment by sagikazarmark** [comment]
As far as I can see the problem is in RingPHP, which is not used in Guzzle 6. Are you sure this is Guzzle 6 report?

Generally speaking: the obvious reason the might cause delays is that a request takes too long. I don't know if newrelic provides a way to profile network operations separately, but if not, I guess it's shown together with the code that executes them. I know that the paid version of Blackfire provides a way to show network operations separately as well, which might be helpful here to decide where the problem is.

Haven't checked the code yet, but I have a strong guess that it has `curl_exec` in it.

To sum up: it's easier IMO to test whether network IO is the problem or something else, so I would start with that.

**Comment by cpats007** [comment]
Thanks for the feedback @sagikazarmark - it's definitely Guzzle 6 - this is my composer entry:

`"guzzlehttp/guzzle": "6.2.0"`

Also, I agree it could be slow network requests but they have a timeout of like 2 seconds on them I think, maybe less - I'll double check but it's set that low as we need responses ideally sub 500ms before we move on. For background, this is part of a `pool` which I have created and sends at once, not sure whether that would make a difference or not over another method of calling an async request?

Would it help to show the code I'm using to build and send the requests?

Thanks again

**Comment by sagikazarmark** [comment]
To be honest we already had some issues like this, but I couldn't really narrow down the problem.

You can see my reproduction environment here: https://github.com/reproduce/guzzle/tree/issue1625

Maybe you could give it a go as well.

As for timeout: can you please check the connect timeout as well? The documentation has some incorrect values, this PR might give a little bit more context: https://github.com/guzzle/guzzle/pull/1524

Last, but not least: can you please check the exact version locked in composer.lock?

**Comment by cpats007** [comment]
Sorry for the delay @sagikazarmark - was away. I'll take a look at your reproduction cases and see what I can come up with.

With regards to the other comments, my `composer.lock` values are showing as follows:

```
"name": "guzzlehttp/guzzle",
"version": "6.2.0",
```

Also, I am setting the timeout in the pool creation options settings as so (slightly sanitised):

```php
$poolConfig = [
            'concurrency' => 10,
            'options'     => [
                'connect_timeout' => 1
            ],
            'fulfilled'   => function ($response, $index) {
                //...
            },
            'rejected'    => function ($response, $index) {
                //...
            },
        ];
```

and then...

```php
$pool = new Pool($this->getClient(), $requests, $poolConfig);
```

I'll have a look into the config settings regarding the timeouts etc - is there a way I can check with the current timeout is set to when initialising a pool or something so I can check if the value is being set accordingly?

Thanks for all the support

**Comment by sagikazarmark** [comment]
I am still totally confused how RingPHP is involved in Guzzle 6 code. I installed the latest Guzzle 6 library and ran this:

`cat composer.lock | grep ring`

No results for ring.

> I can check if the value is being set accordingly?

The linked issue is rather related to defaults than values set improperly, but sure, you can do some simple tests yourself:

1. Connect to localhost on a port where nothing is listening
2. Create a simple script which sleeps for X seconds and set the timeout to X/2

I guess these should prove that the timeouts are working.

I would still suggest some profiling which shows the network overhead of running this code as well. Also, it would be nice to have a clean reproduction environment.

**Comment by cpats007** [comment]
Weird, I have literally deleted it, and reinstalled and this is still what I get:

```
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
  - Installing guzzlehttp/psr7 (1.3.1)
    Downloading: 100%         

  - Installing guzzlehttp/promises (v1.3.1)
    Loading from cache

  - Installing guzzlehttp/guzzle (6.2.0)
    Downloading: 100%         

  - Installing guzzlehttp/streams (3.0.0)
    Loading from cache

  - Installing guzzlehttp/ringphp (1.1.0)
    Loading from cache
```

So it appears it is still there. For now, what I have done is manually set both the `connect_timeout` and `timeout` values and tested that they work and have an effect which they seem to. Hopefully this will overwrite the defaults and prevent any possible long running processes, but it still seems weird, the whole Ring thing.

I'll try and update to the latest version and see if the dependancy disappears

**Comment by cpats007** [comment]
So I just cleared the composer cache, and reinstalled v 6.2.2 and still the same:

```
Installing dependencies (including require-dev) from lock file
  - Installing guzzlehttp/psr7 (1.3.1)
    Downloading: 100%         

  - Installing guzzlehttp/promises (v1.3.1)
    Downloading: 100%         

  - Installing guzzlehttp/guzzle (6.2.2)
    Downloading: 100%         

  - Installing guzzlehttp/streams (3.0.0)
    Downloading: 100%         

  - Installing guzzlehttp/ringphp (1.1.0)
    Downloading: 100%      
```

**Comment by cpats007** [comment]
Okay, step forward - it turns out it was ElasticSearch which was requiring RingPHP and not Guzzle!! I'm looking further into this now, and this may also explain why services were failing when we were having huge traffic spikes...

**Comment by sagikazarmark** [comment]
Okay, so this is not an issue then. Can you please open an issue in ringphp repo instead if you find something there? Thanks

**Comment by mkveksas** [comment]
Hi all,

I believe this issue can be closed. Thanks

**Comment by gregavola** [comment]
@cpats007 I know this s 2+ years ago, did you ever solve this issue?

**Comment by renato-bylyk** [comment]
@cpats007 i know this is 3+ years after @gregavola but did you ever solve this?

**Comment by macpatel** [comment]
ok after 9 years, yes the culprit is curl only. not sure how to resolve it though. As when i pass a handler (curl handler) to my new client, it slows it down. like make it really slow.


EDIT: now instead of default curlhandler i used curlmultihandler and it is running fast now.
